From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: zfcp: deadlock between scheduled task and ERP
References: bnc#487755,LTC#52846

Symptom:     I/O stall.
Problem:     Waiting for the ERP in a task running in global wq
             is a bad idea especially if the ERP needs to run another
             job in this wq before it can finish. ->deadlock.
Solution:    Remove the necessity to wait for the ERP.

Acked-by: John Jolly <jjolly@suse.de>

---
 drivers/s390/scsi/zfcp_erp.c |    2 +-
 drivers/s390/scsi/zfcp_fc.c  |    1 -
 2 files changed, 1 insertion(+), 2 deletions(-)

Index: linux-sles11/drivers/s390/scsi/zfcp_erp.c
===================================================================
--- linux-sles11.orig/drivers/s390/scsi/zfcp_erp.c
+++ linux-sles11/drivers/s390/scsi/zfcp_erp.c
@@ -725,7 +725,6 @@ static int zfcp_erp_adapter_strategy_gen
 		goto failed_openfcp;
 
 	atomic_set_mask(ZFCP_STATUS_COMMON_OPEN, &act->adapter->status);
-	schedule_work(&act->adapter->scan_work);
 
 	return ZFCP_ERP_SUCCEEDED;
 
@@ -1215,6 +1214,7 @@ static void zfcp_erp_action_cleanup(stru
 		break;
 
 	case ZFCP_ERP_ACTION_REOPEN_ADAPTER:
+		schedule_work(&adapter->scan_work);
 		zfcp_adapter_put(adapter);
 		break;
 	}
Index: linux-sles11/drivers/s390/scsi/zfcp_fc.c
===================================================================
--- linux-sles11.orig/drivers/s390/scsi/zfcp_fc.c
+++ linux-sles11/drivers/s390/scsi/zfcp_fc.c
@@ -631,7 +631,6 @@ int zfcp_scan_ports(struct zfcp_adapter 
 	max_entries = chain ? ZFCP_GPN_FT_MAX_ENTRIES : ZFCP_GPN_FT_ENTRIES;
 	max_bytes = chain ? ZFCP_GPN_FT_MAX_SIZE : ZFCP_CT_SIZE_ONE_PAGE;
 
-	zfcp_erp_wait(adapter); /* wait until adapter is finished with ERP */
 	if (fc_host_port_type(adapter->scsi_host) != FC_PORTTYPE_NPORT)
 		return 0;
 
